File: Editorial/0139_Word_Break/Go/3_Bottom_Up_Dynamic_Programming.go
func wordBreak(s string, wordDict []string) bool {
    dp := make([]bool, len(s))
    for i := range s {
        for _, word := range wordDict {
            // Handle out of bounds case
            if i < len(word)-1 {
                continue
            }
            if i == len(word)-1 || dp[i-len(word)] {
                if s[i-len(word)+1:i+1] == word {
                    dp[i] = true
                    break
                }
            }
        }
    }
    return dp[len(s)-1]
}

File: Editorial/0139_Word_Break/JavaScript/3_Bottom_Up_Dynamic_Programming.js
var wordBreak = function (s, wordDict) {
    let dp = new Array(s.length).fill(false);
    for (let i = 0; i < s.length; i++) {
        for (let word of wordDict) {
            // Handle out of bounds case
            if (i < word.length - 1) {
                continue;
            }
            if (i == word.length - 1 || dp[i - word.length]) {
                if (s.substring(i - word.length + 1, i + 1) == word) {
                    dp[i] = true;
                    break;
                }
            }
        }
    }
    return dp[s.length - 1];
};
